MinHeap.java

```diff
@@ -0,0 +1,99 @@
+will move up till it finds its correct place. For ExtractMin, the root will be popped, its value gets replace with the last array value or the
+last leaf to maintain the complete binary tree property. The value gets adjusted by Heapify that searches for the next smallest value 
+that is selected by comparing the parent value with its children as we move down the tree to leaf. If we reach leaf, we replace the root value with
+the found value. 
+*/
+class MinHeap{
+        private int maxsize;
+        private int[] Heap;
+        private int currentSize;
+        MinHeap(int size){
+            this.maxsize = size;
+            this.currentSize = 0;
+            Heap = new int[this.maxsize];           // Creating the Heap Array
+        }
+        int i = -1;
+        public void insert(int data){
+            i++;
+            currentSize = i;
+            Heap[i] = data;
+            int parent = (i-1)/2;               // Searching for the parent of the child since index started is 0.
+            while(parent !=-1 && Heap[parent] > Heap[i]){
+                int t = Heap[parent];
+                Heap[parent] = Heap[i];                 // Swap the parent if the parent value is larger than the child
+                Heap[i] = t;
+                i = parent;
+                parent = (parent-1)/2;
+            }
+            i = currentSize;                        // update the i to the current size of the heap
+        }
+        public int extractMin(){
+            int popped = Heap[0];
+            Heap[0] = Heap[currentSize--];              // Swap the first element/root with the last element of the heap
+            heapify(0);                             // Adjust the heap by heapify (compare the value from root to leaf)
+            return popped;
+        }
+        private boolean isLeaf(int pos) 
+        { 
+            if (pos >= (currentSize / 2) && pos <= currentSize) {       // If leaf node is reached we are done comparing
+                return true; 
+            } 
+            return false; 
+        } 
+        public int getMin(){
+            return Heap[0];                 // Root will carry the least element always
+        }
+        
+        public void heapify(int pos){
+            int left = 2*pos+1;
+            int right = 2*pos+2;
+            if(isLeaf(pos)){
+                return;
+            }
+            if((Heap[pos] > Heap[left]) || (Heap[pos] > Heap[right])){
+                if(Heap[left] > Heap[right]){
+                int t = Heap[pos];
+                Heap[pos] = Heap[right];                // Compare the parent with children and replace the smallest child with parent
+                Heap[right] = t;
+                } else {
+                int t = Heap[pos];
+                Heap[pos] = Heap[left];
+                Heap[left] = t;
+                }
+                heapify(pos);                       // Continue heapify till the heap values are completely adjusted
```

Comment:
Correct

Reply:
What is the time complexity of heapify function alone?

Reply:
The heapify function gives O(n) time complexity as per my read. I have a question here that I am trying to understand how the heap building takes O(n). Now, consider I am deleting an element. I am calling heapify once I delete and element, and then heapify takes over to adjust the heap in O(log n). So total deletions are n elements and for each element, I call heapify 'n' times and it adjusts the heap in O(log n) => complexity O(nlogn) (similar to heap sort).  But I am not able to understand why do we say heapify takes O(n) since we are comparing the top element with children and moving down to leaf which takes log n steps right? I have been referring to this article: https://stackoverflow.com/questions/9755721/how-can-building-a-heap-be-on-time-complexity but not able to understand the logic behind O(n) behind heapify (ignore the maths part as of now). Can you please give me a brief explanation to my question?